File: solutions/_234_palindrome_linked_list/Solution.java
package _234_palindrome_linked_list;

class Solution {

  public boolean isPalindrome(ListNode head) {
    ListNode fast = head;
    ListNode slow = head;

    // find the middle node (slow)
    while (fast != null && fast.next != null) {
      fast = fast.next.next;
      slow = slow.next;
    }

    // reverse second half
    ListNode left = head;
    ListNode right = reverse(slow);

    // check for palindrome
    while (right != null && left != null) {
      if (right.val != left.val) return false;
      right = right.next;
      left = left.next;
    }

    return true;
  }

  private ListNode reverse(ListNode cur) {
    ListNode prev = null;
    while (cur != null) {
      ListNode tmp = cur.next;
      cur.next = prev;
      prev = cur;
      cur = tmp;
    }
    return prev;
  }
}
/*
https://leetcode.com/problems/palindrome-linked-list/
*/

File: solutions/_234_palindrome_linked_list/solution.py

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        fast = head
        slow = head
        
        # find the middle (slow)
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            
        # reverse second half
        left = head
        right = self.reverse(slow)
        
        # check palindrome
        while right:
            if left.val != right.val:
                return False
            left = left.next
            right = right.next
        return True
    
    def reverse(self, cur: ListNode) -> ListNode:
        prev = None
        while cur:
            tmp = cur.next
            cur.next = prev
            prev = cur
            cur = tmp
        return prev
        
  
# https://leetcode.com/problems/palindrome-linked-list/
